libraries/tock-cells/src/map_cell.rs

```diff
@@ -12,36 +13,22 @@ use core::{mem, ptr};
 /// `Option` wrapped in a `RefCell` --- attempts to take the value from inside a
 /// `MapCell` may fail by returning `None`.
 pub struct MapCell<T> {
-    val: UnsafeCell<U<T>>,
+    val: UnsafeCell<MaybeUninit<T>>,
     occupied: Cell<bool>,
 }
 
-// This allows us to mimic `mem::uninitialized` in a way that can be marked `const`.
-//
-// Specifically, we just want to allocate some memory the size of some
-// particular `T` and we don't care what's there---this happens to not be
-// marked `const` in the core library right now since it's an LLVM intrinsic.
-//
-// The `occupied` field of the MapCell tracks whether it is valid to access the
-// `some` variant of this union.
-#[allow(unions_with_drop_fields)]
-union U<T> {
-    none: (),
-    some: T,
-}
-
 impl<T> MapCell<T> {
     pub const fn empty() -> MapCell<T> {
         MapCell {
-            val: UnsafeCell::new(U { none: () }),
+            val: UnsafeCell::new(MaybeUninit::uninit()),
```

Comment:
This is defined behavior and is fine, though "unsafe" and we need to make sure we don't `assume_init` or `get` this value until it's been initialized.

Reply:
Agreed, I think since the value is private we can leave it out of the docs, but I'll add a comment.